Various testutil improvements

thaJeztah opened the pull request:
See individual commits for a description of the changes

thaJeztah committed:
hack/make: don't attempt to unmount non-existing daemon root-dir

Before:

    DONE 2 tests in 12.272s
    ---> Making bundle: .integration-daemon-stop (in bundles/test-integration)
    umount: bundles/test-integration/root: mountpoint not found

After:

    DONE 2 tests in 14.650s
    ---> Making bundle: .integration-daemon-stop (in bundles/test-integration)

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
testutil/daemon: daemon.Cleanup(): unmount daemon root-dir as part of cleanup

test-daemons remove their docker.pid when stopped, so the `.integration-daemon-stop`
script did not find the mounts for those daemons, and therefore was not unmounting
them.

As a result, cleaning up the bundles directory on consecutive runs of the tests would fail;

    rm: cannot remove 'bundles/test-integration/TestDockerSwarmSuite/TestSwarmInit/d1f188f3f5472/root': Device or resource busy

This patch unmounts the root directory of the daemon as part of the cleanup step.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
testutil/daemon: prefix all logs with daemon-id

This makes it easier to debug issues with tests that
start multiple daemons.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
testutil/daemon: wrap errors

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
testutil/daemon: store pidfile-path, and ignore errors when removing

This patch stores the location of the pidfile, so that we can use the
same path that was set to create it. If no pidfile was created, we'll
not try to remove it.

We're now also ignoring errors when removing the pidfile, as they should
not fail the test (especialy if no pidfile was created in the first place,
as that could potentially hide the actual failure).

This may help with "failures" such as the one below:

```
FAIL: check_test.go:347: DockerSwarmSuite.TearDownTest

check_test.go:352:
    d.Stop(c)
/go/src/github.com/docker/docker/internal/test/daemon/daemon.go:414:
    t.Fatalf("Error while stopping the daemon %s : %v", d.id, err)
... Error: Error while stopping the daemon d1512c423813a : remove /go/src/github.com/docker/docker/bundles/test-integration/DockerSwarmSuite.TestServiceLogs/d1512c423813a/docker.pid: no such file or directory
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
testutil/daemon: print all arguments when failing to start daemon

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
testutil/daemon: always remove pidfile after daemon is stopped

If the daemon was stopped successfully in one of the retry-loops,
the function would return early;

```go
for {
	select {
	case err := <-d.Wait:
---> the function returns here, both on "success" and on "fail"
		return err
	case <-time.After(20 * time.Second):
...
```

In that case, the pidfile would not be cleaned up. This patch changes
the function to clean-up the pidfile in a defer, so that it will
always be removed after succesfully stopping the daemon.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
ping @tiborvass @SamWhited @kolyshkin ptal

thaJeztah commented on hack/make/.integration-daemon-stop:
```diff
@@ -11,7 +11,9 @@ if [ ! "$(go env GOOS)" = 'windows' ]; then
 		root=$(dirname "$pidFile")/root
-		umount "$root" || true
+		if [ -d "$root" ]; then
+			umount "$root" || true
```
in practice, this only finds the `docker.pid` of the daemon that was started by `.integration-daemon-start`, but that daemon uses the default (`/var/lib/docker`) as path, so nothing gets unmounted. I think this will only do something if a daemon was not killed as part of a test

kolyshkin replied:
I'd rephrase as
```shell
test -d "$root" && umount "$root" || true
```

kolyshkin replied:
and even then you can still get "not mounted" error (afaik there's no way to work around it)

thaJeztah replied:
Ah, yes, that'd work

I realize it won't fix all situations (always gonna be a race), but currently, it _always_ prints this message

thaJeztah replied:
hm, so shellcheck doesn't like your approach, and would require a `//nolint` for that; https://github.com/koalaman/shellcheck/wiki/SC2015

I'll keep the current approach (although it's slightly more verbose)

kolyshkin replied:
Ahh, this is the case when shellcheck is making things worse (and yes, I would not want to add `// nolint`).

Let it stay as it is.

thaJeztah commented:
Whoops; made a typo; fixed (also found some more small changes)

```
testutil/daemon/daemon.go:95: printf: Wrapf format % is missing verb at end of string (govet)
  		return nil, errors.Wrapf(err, "failed to create daemon socket root %", SockRoot)
```

thaJeztah commented on testutil/daemon/daemon.go:
```diff
@@ -382,6 +389,9 @@ func (d *Daemon) StartWithBusybox(t testing.TB, arg ...string) {
 func (d *Daemon) Kill() error {
 	if d.cmd == nil || d.Wait == nil {
+		if err := mount.Unmount(d.Root); err != nil {
+			d.log.Logf("[%s] failed to unmount daemon root (%s): %v", d.id, d.Root, err)
```
Dropped this commit, because this would also unmount the daemon root if live-restore is enabled, probably causing this failure:

```
 === FAIL: amd64.integration-cli TestDockerDaemonSuite/TestExecWithUserAfterLiveRestore (2.90s)
     --- FAIL: TestDockerDaemonSuite/TestExecWithUserAfterLiveRestore (2.90s)
         daemon.go:26: Creating a new daemon at: "/go/src/github.com/docker/docker/bundles/test-integration/TestDockerDaemonSuite/TestExecWithUserAfterLiveRestore"
         docker_cli_daemon_test.go:2723: assertion failed: expression is false: err == nil: Output: unable to find user test: no matching entries in passwd file
```

kolyshkin commented on testutil/daemon/daemon_unix.go:
```diff
@@ -13,17 +13,17 @@ import (
-	netnsPath := filepath.Join(execRoot, "netns")
+	netnsPath := filepath.Join(d.execRoot, "netns")
 	filepath.Walk(netnsPath, func(path string, info os.FileInfo, err error) error {
 		if err := unix.Unmount(path, unix.MNT_DETACH); err != nil && err != unix.EINVAL && err != unix.ENOENT {
```
While at it, we can switch to using `mount.Unmount()` which does all the same things.

Yet better, use `RecursiveUnmount()` to further simplify things here.

thaJeztah replied:
I now recall I had a PR that made changes here; https://github.com/moby/moby/pull/36511, but for some reason it kept failing.

Let me try using the RecursiveUnmount() that sounds like a clean approach (if it works)

kolyshkin replied:
I take that back. Neither `filepath.Walk()` nor `RecursiveUnmount()` should be used here, as per https://github.com/moby/moby/pull/36511#discussion_r172975216

kolyshkin replied:
I.e. the approach in #36511 is the correct one; I would take that one here )

thaJeztah replied:
Let me rebase the other one, to keep this one simpler (and in case it's still problematic). I can rebase it on top of this one

kolyshkin commented on testutil/daemon/daemon.go:
```diff
@@ -246,11 +247,15 @@ func (d *Daemon) StartWithLogFile(out *os.File, providedArgs ...string) error {
 	}
 
+	if d.pidFile == "" {
+		d.pidFile = fmt.Sprintf("%s/docker.pid", d.Folder)
```
To me, `path.Join(d.Folder, "docker.pid")` looks more elegant than this `Sprintf`.

thaJeztah replied:
🤦‍♂ yes, should've seen that 👍

kolyshkin commented on testutil/daemon/daemon.go:
```diff
@@ -395,7 +400,10 @@ func (d *Daemon) Kill() error {
 	}
 
-	return os.Remove(fmt.Sprintf("%s/docker.pid", d.Folder))
+	if d.pidFile != "" {
```
Would be good to note in the commit message that this fixes bogus error from removing `docker.pid` that was never created.

kolyshkin replied:
@SamWhited I would 
- not care about error returned from `os.Remove()` at all
- not return any error here as this will fail the test

I.e. current approach looks fine for me.

kolyshkin commented on testutil/daemon/daemon.go:
```diff
@@ -519,12 +524,7 @@ out2:
-		return os.Remove(d.pidFile)
-	}
-	return nil
+	return d.cmd.Wait()
```
Not quite related to this patch (except see below), but I see that `StartWithLogFile()` already creates a goroutine that calls `d.cmd.Waits()` and sends its return to the channel, `d.Wait`.

Now, this code calls `d.cmd.Wait()` one more time, I wonder why, one wait should be just enough. Looking into git logs, it comes all the way from ancient history. Maybe it's time to fix it. I would remove this second wait.

Now that you're returning the second `d.cmd.Wait()` exit code, it might actually return an error (I haven't checked that, just a hunch) as it was already been waited for.

thaJeztah replied:
done

kolyshkin reviewed (changes requested):
left a few nitpicks and one important comment about double `d.cmd.Wait()`

thaJeztah committed:
testutil/daemon: remove redundant d.cmd.Wait()

`daemon.StartWithLogFile()` already creates a goroutine that
calls `d.cmd.Waits()` and sends its return to the channel, `d.Wait`.

This code called `d.cmd.Wait()` one more time, and returns the
error, which may produce an error _because_ it's called a second
time, and potentially cause an incorrect test-result.

(thanks to Kir Kolyshkin for spotting this)

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

SamWhited commented on testutil/daemon/daemon.go:
```diff
@@ -451,12 +461,17 @@ func (d *Daemon) StopWithError() (err error) {
-			d.log.Logf("[%s] Error when stopping daemon: %v", d.id, err)
+			d.log.Logf("[%s] daemon stopped", d.id)
+			if d.pidFile != "" {
+				_ = os.Remove(d.pidFile)
```
It could potentially let us know that something is wrong with our tests. I'd say that it's certainly not worth failing tests over, but is worth logging personally, but I defer to your judgement.

kolyshkin commented on testutil/daemon/daemon.go:
```diff
@@ -231,7 +234,7 @@ func (d *Daemon) Start(t testing.TB, args ...string) {
 	logFile, err := os.OpenFile(filepath.Join(d.Folder, "docker.log"), os.O_RDWR|os.O_CREATE|os.O_APPEND, 0600)
 	if err != nil {
-		return errors.Wrapf(err, "[%s] Could not create %s/docker.log", d.id, d.Folder)
+		return errors.Wrapf(err, "[%s] failed to create %s/docker.log", d.id, d.Folder)
```
Actually, `os.OpenFile()` should have the file name in its error message, so printing it explictly is redundant. Just
```go
errors.Wrap(err, "[%s] failed to create logfile")
```
should be enough here.

thaJeztah replied:
Good point, yes I can remove it

thaJeztah replied:
done

kolyshkin reviewed (approved):
LGTM

thaJeztah commented:
Argh; Windows failed, but for some reason, Windows RS5 never prints the actual failure

```

[2019-10-10T18:20:14.730Z] ERROR: make.ps1 failed:
[2019-10-10T18:20:14.730Z] Unit tests failed
[2019-10-10T18:20:14.730Z] At C:\gopath\src\github.com\docker\docker\hack\make.ps1:324 char:32
[2019-10-10T18:20:14.730Z] +     if ($LASTEXITCODE -ne 0) { Throw "Unit tests failed" }
[2019-10-10T18:20:14.730Z] +                                ~~~~~~~~~~~~~~~~~~~~~~~~~
[2019-10-10T18:20:14.730Z] 
```

It could be a compile error, and if I remember correctly, Windows RS1 **does** print the actual failure (using the exact same script 🤷‍♂); I'm gonna trigger a rebuild with RS1 enabled to see if I can find what's wrong.

Probably also open a ticket for it, so that someone with more Windows knowledge could help fixing that

thaJeztah commented:
There you go; Windows RS1 correctly shows the compile error, whereas RS5 just hides it (and both report all unit tests to have "passed", but possibly the problem occurs after that);

```
testutil\daemon\daemon.go:221:25: cannot use d (type *Daemon) as type string in argument to cleanupNetworkNamespace
```

thaJeztah commented:
all green now

@cpuguy83 @tiborvass PTAL

tiborvass reviewed (approved):
LGTM but we should probably also add more `t.Helper()` in `testutil/daemon` (when daemons are starting/stopping the file lines are completely useless)

cpuguy83 reviewed (approved):
LGTM

thaJeztah commented:
> but we should probably also add more t.Helper() in testutil/daemon (when daemons are starting/stopping the file lines are completely useless)

agreed; I didn't go looking for other ones yes; this one just stood out while testing my changes
